cloudevents/event_v02.go

```diff
@@ -0,0 +1,260 @@
+	}
+	if ec.SchemaURL != "" {
+		h.Set("CE-SchemaURL", ec.SchemaURL)
+	}
+	if ec.ContentType != "" {
+		h.Set("Content-Type", ec.ContentType)
+	}
+	for k, v := range ec.Extensions {
+		// Per spec, map-valued extensions are converted to a list of headers as:
+		// CE-attrib-key
+		if mapVal, ok := v.(map[string]interface{}); ok {
+			for subkey, subval := range mapVal {
+				data := fmt.Sprint(subval)
+				encoded, err := json.Marshal(subval)
+				if err != nil {
+					data = string(encoded)
+				}
+				h.Set("CE-"+k+"-"+subkey, data)
+			}
+			continue
+		}
+		data := fmt.Sprint(v)
+		encoded, err := json.Marshal(v)
+		if err != nil {
+			data = string(encoded)
+		}
+		h.Set("CE-"+k, data)
+	}
+
+	return h
+}
+
+// FromHeaders implements the BinaryLoader interface.
+func (ec *V02EventContext) FromHeaders(in http.Header) error {
+	missingField := func(name string) error {
+		if in.Get("CE-"+name) == "" {
+			return fmt.Errorf("Missing field %q in %v: %q", "CE-"+name, in, in.Get("CE-"+name))
+		}
+		return nil
+	}
+	err := anyError(
+		missingField("SpecVersion"),
+		missingField("ID"),
+		missingField("Type"),
+		missingField("Source"),
+	)
+	if err != nil {
+		return err
+	}
+	data := V02EventContext{
+		ContentType: in.Get("Content-Type"),
+		Extensions:  make(map[string]interface{}),
+	}
+	// Extensions and top-level fields are mixed under "CE-" headers.
+	// Extract them all here rather than trying to clear fields in headers.
+	for k, v := range in {
+		if strings.EqualFold(k[:len("CE-")], "CE-") {
+			key := strings.ToLower(string(k[len("CE-"):]))
+			value := v[0]
+			switch key {
+			case "specversion":
+				data.SpecVersion = value
+			case "type":
+				data.Type = value
+			case "source":
+				data.Source = value
+			case "id":
+				data.ID = value
+			case "schemaurl":
+				data.SchemaURL = value
+			case "time":
+				if data.Time, err = time.Parse(time.RFC3339Nano, value); err != nil {
+					return err
+				}
+			default:
+				var tmp interface{}
+				if err = json.Unmarshal([]byte(value), &tmp); err != nil {
+					tmp = value
+				}
+				// Per spec, map-valued extensions are converted to a list of headers as:
+				// CE-attrib-key. This is where things get a bit crazy... see
+				// https://github.com/cloudevents/spec/issues/367 for additional notes.
+				if strings.Contains(key, "-") {
+					items := strings.SplitN(k, "-", 2)
+					key, subkey := items[0], items[1]
+					if _, ok := data.Extensions[key]; !ok {
+						data.Extensions[key] = make(map[string]interface{})
+					}
+					if submap, ok := data.Extensions[key].(map[string]interface{}); ok {
+						submap[subkey] = tmp
+					}
+				} else {
+					data.Extensions[key] = tmp
+				}
+			}
+		}
+	}
+	*ec = data
+	return nil
+}
+
+// AsJSON implementsn the StructuredSender interface.
+func (ec V02EventContext) AsJSON() (map[string]json.RawMessage, error) {
+	ret := make(map[string]json.RawMessage)
+	err := anyError(
+		encodeKey(&ret, "specversion", ec.SpecVersion),
```

Comment:
this function is elsewhere, but maps are passed by reference and there is no need to take pass around pointer to it.